controllers/topology/internal/extensions/patches/engine.go

```diff
@@ -0,0 +1,495 @@
+	req := &api.GenerateRequest{}
+
+	// Calculate global variables.
+	globalVariables, err := variables.Global(blueprint.Topology, desired.Cluster)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate global variables")
+	}
+	req.Variables = globalVariables
+
+	// Add the InfrastructureClusterTemplate.
+	t, err := newTemplate(blueprint.InfrastructureClusterTemplate, api.TemplateID{
+		APIVersion: blueprint.InfrastructureClusterTemplate.GetAPIVersion(),
+		Kind:       blueprint.InfrastructureClusterTemplate.GetKind(),
+		TargetType: api.InfrastructureClusterTemplateTargetType,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create InfrastructureCluster template %s",
+			tlog.KObj{Obj: blueprint.InfrastructureClusterTemplate})
+	}
+	req.Items = append(req.Items, t)
+
+	// Calculate controlPlane variables.
+	controlPlaneVariables, err := variables.ControlPlane(&blueprint.Topology.ControlPlane, desired.ControlPlane.Object)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate ControlPlane variables")
+	}
+
+	// Add the ControlPlaneTemplate.
+	t, err = newTemplate(blueprint.ControlPlane.Template, api.TemplateID{
+		APIVersion: blueprint.ControlPlane.Template.GetAPIVersion(),
+		Kind:       blueprint.ControlPlane.Template.GetKind(),
+		TargetType: api.ControlPlaneTemplateTargetType,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create ControlPlane template %s",
+			tlog.KObj{Obj: blueprint.ControlPlane.Template})
+	}
+	t.Variables = controlPlaneVariables
+	req.Items = append(req.Items, t)
+
+	// If the clusterClass mandates the controlPlane has infrastructureMachines,
+	// add the InfrastructureMachineTemplate for control plane machines.
+	if blueprint.HasControlPlaneInfrastructureMachine() {
+		t, err = newTemplate(blueprint.ControlPlane.InfrastructureMachineTemplate, api.TemplateID{
+			APIVersion: blueprint.ControlPlane.InfrastructureMachineTemplate.GetAPIVersion(),
+			Kind:       blueprint.ControlPlane.InfrastructureMachineTemplate.GetKind(),
+			TargetType: api.ControlPlaneInfrastructureMachineTemplateTargetType,
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create ControlPlane's machine template %s",
+				tlog.KObj{Obj: blueprint.ControlPlane.InfrastructureMachineTemplate})
+		}
+		t.Variables = controlPlaneVariables
+		req.Items = append(req.Items, t)
+	}
+
+	// Add BootstrapConfigTemplate and InfrastructureMachine template for all MachineDeploymentTopologies
+	// in the Cluster.
+	// NOTE: We intentionally iterate over MachineDeploymentTopologies in the Cluster instead of over
+	// MachineDeploymentClasses in the ClusterClass. We have to calculate patches per MachineDeploymentTopology
+	// instead of per MachineDeploymentClass, because the builtin variable values depend on the current
+	// MachineDeployment. E.g., there are variables for the version or replicas of the MachineDeployment.
```

Comment:
```suggestion
	// NOTE: We intentionally iterate over MachineDeployment in the Cluster instead of over
	// MachineDeploymentClasses in the ClusterClass because each MachineDeployment in a topology
	// has its own state, e.g. version or replicas, and it is required to use those info as built in variable 
	// e.g. to compute the machine image for a specific Kubernetes version.
```